### PR TITLE
Fix issue with jersey and unix domain sockets

### PR DESCRIPTION
**Pull request by marcuslinke**
This eliminates the socket already closed exception (org.newsclub.net.unix.AFUNIXSocketException: Bad file descriptor). The problem occurs when the underlying connection of a stream is closed and chunked encoding is active. The apache `ChunkedInputStream` still tries to read till the end of the current chunk while `close()`ing itself. This is not possible as the connection/socket is already closed.

Sadly it needs to fix two classes `AFUNIXSocketImpl` and  `ChunkedInputStream` to solve the issue.

This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/697)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -115,11 +116,20 @@
...
+			<groupId>com.kohlschutter.junixsocket</groupId>
+			<artifactId>junixsocket-common</artifactId>
+			<version>${junixsocket.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>com.kohlschutter.junixsocket</groupId>
+			<artifactId>junixsocket-native-common</artifactId>
+			<version>${junixsocket.version}</version>
+		</dependency>
+		<dependency>
```
Comment (marcuslinke) [inline pom.xml]: Dependency must be removed...

File: pom.xml [hunk 2]
```diff
@@ -266,13 +271,13 @@
...
 	<build>
-<!-- 		<extensions> -->
-<!-- 			<extension> -->
-<!-- 				<groupId>kr.motd.maven</groupId> -->
-<!-- 				<artifactId>os-maven-plugin</artifactId> -->
-<!-- 				<version>1.2.3.Final</version> -->
-<!-- 			</extension> -->
-<!-- 		</extensions> -->
+		<!-- <extensions> -->
+		<!-- <extension> -->
```
Comment (KostyaSha) [inline pom.xml]: delete it?

**Comment by KostyaSha** [comment]
👍

File: pom.xml (hunk 2 above)
Comment (marcuslinke) [inline pom.xml]: Makes sense 👍

**Comment by marcuslinke** [comment]
@KostyaSha What is the difference between the two travis builds `continuous-integration/travis-ci/push`and `continuous-integration/travis-ci/pr` ?

**Comment by KostyaSha** [comment]
Obviously pull request and branch build ;)

**Comment by KostyaSha** [comment]
So if you don't plan cooperate on branch, then do branch in your fork and open PR.

**Comment by KostyaSha** [comment]
```
(com.github.dockerjava.core.command.EventsCmdImplTest)  Time elapsed: 29.805 sec  <<< FAILURE!
java.lang.AssertionError: Received only: [] expected [true] but found [false]
	at com.github.dockerjava.core.command.EventsCmdImplTest.testEventStreamingWithFilter(EventsCmdImplTest.java:122)
Results :
Failed tests: 
com.github.dockerjava.core.command.EventsCmdImplTest.(com.github.dockerjava.core.command.EventsCmdImplTest)
  Run 1: PASS
  Run 2: PASS
  Run 3: PASS
  Run 4: EventsCmdImplTest.testEventStreamingWithFilter:122->Assert.assertTrue:42->Assert.failNotEquals:513->Assert.fail:94 Received only: [] expected [true] but found [false]

``` 
retriggering build

**Comment by KostyaSha** [comment]
Also place somewhere in javadoc why clases was copied.

**Comment by KostyaSha** [comment]
@marcuslinke please set milestones before merge and update changelog after merge

**Comment by KostyaSha** [comment]
```
[WARNING] docker-java-3.0.6.jar, httpcore-4.4.5.jar define 1 overlapping classes: 
[WARNING]   - org.apache.http.impl.io.ChunkedInputStream
[WARNING] docker-java-3.0.6.jar, junixsocket-common-2.0.4.jar define 5 overlapping classes: 
[WARNING]   - org.newsclub.net.unix.AFUNIXSocketImpl$AFUNIXOutputStream
[WARNING]   - org.newsclub.net.unix.AFUNIXSocketImpl
[WARNING]   - org.newsclub.net.unix.AFUNIXSocketImpl$AFUNIXInputStream
[WARNING]   - org.newsclub.net.unix.AFUNIXSocketImpl$Lenient
[WARNING]   - org.newsclub.net.unix.AFUNIXSocketImpl$1
```

**Comment by KostyaSha** [comment]
@marcuslinke where ChunkedInputStream is used? or do you trying replace classes? Classed are messed..

**Comment by KostyaSha** [comment]
On mac after shading i have 
```
Caused by: java.lang.UnsatisfiedLinkError: com.github.kostyasha.yad_docker_java.org.newsclub.net.unix.NativeUnixSocket.connect(Ljava/lang/String;Ljava/io/FileDescriptor;)V
	at com.github.kostyasha.yad_docker_java.org.newsclub.net.unix.NativeUnixSocket.connect(Native Method)
	at com.github.kostyasha.yad_docker_java.org.newsclub.net.unix.AFUNIXSocketImpl.connect(AFUNIXSocketImpl.java:137)
	at com.github.kostyasha.yad_docker_java.org.newsclub.net.unix.AFUNIXSocket.connect(AFUNIXSocket.java:107)
	at com.github.kostyasha.yad_docker_java.com.github.dockerjava.jaxrs.ApacheUnixSocket.connect(ApacheUnixSocket.java:62)
	at com.github.kostyasha.yad_docker_java.com.github.dockerjava.jaxrs.UnixConnectionSocketFactory.connectSocket(UnixConnectionSocketFactory.java:74)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:134)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at com.github.kostyasha.yad_docker_java.org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at com.github.kostyasha.yad_docker_java.org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:435)
	at com.github.kostyasha.yad_docker_java.org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:252)
	... 81 more
```
Will try compare with linux env and plain docker-java.

**Comment by KostyaSha** [comment]
M... it impossible relocate java class because it depends on native library?

**Comment by marcuslinke** [comment]
@KostyaSha Yes, I've replaced two classes as described in the first comment. Didn't know that this is a problem with shading... :(

**Comment by KostyaSha** [comment]
Well, it warning, but having conflicting package-class is questionable.

**Comment by marcuslinke** [comment]
Yes, thats ugly but fixing this in upstream projects (apache HTTPCore AND junixsockets) would take ages I think.

**Comment by papegaaij** [comment]
You cannot simply copy a class an expect it to shadow a class from a dependency. This all depends on classloading order, which can be non-deterministic in some cases (such as packaging multiple jars in a war or ear). Your only options are to get this fixed upstream or to work around this some other way. The currently solution breaks our build because the maven-enforcer-plugin detects duplicate classes.

**Comment by alexburchak** [comment]
My project became a victim of the class loading order issue :)
ChunkedInputStream in docker-java (from the commit), which has that try/catch fix, is not loaded and original class implementation from httpcore is used instead.
Exception thrown in close() method (after invoking read()) is not caught and thus propagated above.

**Pull request merged**